### PR TITLE
[PIRK-25] Serialization and storage changes to Querier, Query, and Response.

### PR DESCRIPTION
**Pull request by tellison**
- Introduce serialization services that can flatten objects into a
variety of different formats.  Currently just reusing the Java
serialization code until the project is ready to adopt other
serializers.
 
 - Introduce storage services that can store objects and retrieve them.
Currently just reusing the local file system and HDFS.  The storage
service uses one of the serialization services to flatten the objects.
 
 - Remove code from Querier, Query, and Response that was serializing
and storing each of them; and update the callers to use the storage
services.

### CODE REVIEW

File: src/main/java/org/apache/pirk/serialization/HadoopFileSystemStore.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+  public void store(String pathName, Storable value) throws IOException
+  {
+    store(new Path(pathName), value);
+  }
+
+  public void store(Path path, Storable obj) throws IOException
+  {
+    OutputStream os = hadoopFileSystem.create(path);
+    try
```
Comment (smarthi) [inline src/main/java/org/apache/pirk/serialization/HadoopFileSystemStore.java]: How bout using a try-all here ?

File: src/main/java/org/apache/pirk/serialization/HadoopFileSystemStore.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+  public <T> T recall(String pathName, Class<T> type) throws IOException
+  {
+    return recall(new Path(pathName), type);
+  }
+
+  public <T> T recall(Path path, Class<T> type) throws IOException
+  {
+    InputStream is = hadoopFileSystem.open(path);
+    try
```
Comment (smarthi) [inline src/main/java/org/apache/pirk/serialization/HadoopFileSystemStore.java]: try-all ? Here and other places in this PR.

**Comment by tellison** [comment]
@smarthi   Sorry, not groking "try-all" can you explain?

These follow the method overloading pattern seen in a variety of places like  Arrays.fill(byte[], byte), Arrays.fill(char[], char), etc.

**Comment by tellison** [comment]
Ha, managed to take down the Travis VM ;-)
> *** buffer overflow detected ***: /usr/lib/jvm/java-7-openjdk-amd64/jre/bin/java terminated

Yet it worked on my Travis branch before I opened the PR.  Maybe an admin can retry it pls.

**Comment by smarthi** [comment]
@tellison meant to say use  - try-all-resources

`//    FileOutputStream fos = new FileOutputStream(file);
    try(FileOutputStream fos = new FileOutputStream(file)){ .... }`

File: src/main/java/org/apache/pirk/serialization/JavaSerializer.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  {
+    ObjectOutputStream oos = new ObjectOutputStream(stream);
+    oos.writeObject(obj);
+  }
+
+  @SuppressWarnings("unchecked")
+  public <T> T read(InputStream stream, Class<T> type) throws IOException
+  {
+    ObjectInputStream oin = new ObjectInputStream(stream);
+    try
```
Comment (smarthi) [inline src/main/java/org/apache/pirk/serialization/JavaSerializer.java]: how bout 
`try(ObjectInputStream oin = new ObjectInputStream(stream);)`

File: src/main/java/org/apache/pirk/serialization/LocalFileSystemStore.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  public void store(String path, Storable obj) throws IOException
+  {
+    store(new File(path), obj);
+  }
+
+  public void store(File file, Storable obj) throws IOException
+  {
+    FileOutputStream fos = new FileOutputStream(file);
+    try
```
Comment (smarthi) [inline src/main/java/org/apache/pirk/serialization/LocalFileSystemStore.java]: `try(FileOutputStream fos = new FileOutputStream(file);)`

**Comment by tellison** [comment]
@smarthi  Ah, ok.  Yep, I had mistakenly thought that try-with-resources was a Java 8 feature :-)  I'll update in due course (not today).

**Comment by ellisonanne** [comment]
Very nice - much cleaner. 

One very minor thing that you may want to go ahead and change in ExpTableMapper:

`query = new HadoopFileSystemStore(FileSystem.newInstance(ctx.getConfiguration())).recall(queryDir, Query.class);`
 
to

`query = new HadoopFileSystemStore(fs).recall(queryDir, Query.class);`

as the FileSystem fs has already been grabbed a couple of lines above.

Otherwise, we will need to make sure that custom StorageServices and SerializationServices have corresponding tests.

**Comment by ellisonanne** [comment]
`fs` is defined in setup of `ExpTableMapper`:

 ```
public void setup(Context ctx) throws IOException, InterruptedException
  {
    super.setup(ctx);

    valueOut = new Text();

    FileSystem fs = FileSystem.newInstance(ctx.getConfiguration());
    String queryDir = ctx.getConfiguration().get("pirMR.queryInputDir");
    query = Query.readFromHDFSFile(new Path(queryDir), fs);

    dataPartitionBitSize = query.getQueryInfo().getDataPartitionBitSize();
    maxValue = (int) Math.pow(2, dataPartitionBitSize) - 1;

    NSquared = query.getNSquared();
  }
```

**Comment by tellison** [comment]
Ah, I guess you are looking at the current code (pre-PR).  I in-lined that ```fs``` definition as it is only used once.

**Comment by ellisonanne** [comment]
Correct - I was looking at the old code

+1

File: src/main/java/org/apache/pirk/serialization/SerializationService.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+{
+  public abstract <T> T read(InputStream stream, Class<T> type) throws IOException;
+
+  public abstract void write(OutputStream w, Storable obj) throws IOException;
+
+  public byte[] toBytes(Storable obj)
+  {
+    ByteArrayOutputStream bos = new ByteArrayOutputStream();
+
+    try
```
Comment (smarthi) [inline src/main/java/org/apache/pirk/serialization/SerializationService.java]: Change to try(ByteArrayOutputStream bos = new ByteArrayOutputStream(); )

**Pull request closed without merging**